### PR TITLE
Add support for printing RenderTree for RemoteFrame(s)

### PR DESCRIPTION
**Pull request by chirags27**
#### 7ce8aed08237330dbdb86c0c944e2b4ff538fb57
<pre>
Add support for printing RenderTree for RemoteFrame(s)
<a href="https://bugs.webkit.org/show_bug.cgi?id=256972">https://bugs.webkit.org/show_bug.cgi?id=256972</a>
rdar://105023551

Reviewed by Alex Christensen.

This change adds infrastructure so that we can get RenderTree dumps for
RemoteFrame&apos;s which are not hosted in the process which gets the request
to print the RenderTreeAsText

* LayoutTests/http/tests/site-isolation/basic-iframe-render-output-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/basic-iframe-render-output.html: Added.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::elementAccessibilityHitTest const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::addAttachmentChildren):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::accessibilityIgnoreAttachment const):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::isFrameElement):
* Source/WebCore/page/FrameView.h:
(isType):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::writeRenderTreeAsText):
* Source/WebCore/page/LocalFrameView.h:
(isType):
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::renderTreeAsText):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebCore/page/RemoteFrameView.cpp:
(WebCore::RemoteFrameView::writeRenderTreeAsText):
* Source/WebCore/page/RemoteFrameView.h:
* Source/WebCore/page/ios/EventHandlerIOS.mm:
(WebCore::EventHandler::passSubframeEventToSubframe):
* Source/WebCore/page/mac/EventHandlerMac.mm:
(WebCore::EventHandler::passSubframeEventToSubframe):
(WebCore::EventHandler::platformCompletePlatformWidgetWheelEvent):
* Source/WebCore/platform/Widget.h:
(WebCore::Widget::isLocalFrameView const):
(WebCore::Widget::isFrameView const): Deleted.
* Source/WebCore/rendering/RenderTreeAsText.cpp:
(WebCore::write):
(WebCore::externalRepresentationForLocalFrame):
* Source/WebCore/rendering/RenderTreeAsText.h:
(WebCore::externalRepresentationForLocalFrame):
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::updateWidgetGeometry):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::renderTreeAsText):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::renderTreeAsText):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::didReceiveSyncMessage):
(WebKit::WebProcess::renderTreeAsText):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/265046@main">https://commits.webkit.org/265046@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f16517f73ff18e0fbca5d5edd7e18cf174b4d778

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9568 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10069 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11229 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9373 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9579 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9784 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12284 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9720 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10585 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8163 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11388 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/9683 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7880 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16119 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8975 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8851 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12213 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9344 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7596 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8543 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8607 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12765 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1104 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9115 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/page/RemoteFrame.cpp [hunk 1]
```diff
@@ -105,4 +105,9 @@ void RemoteFrame::frameDetached()
     m_client->frameDetached();
 }
 
+String RemoteFrame::renderTreeDump(int baseIndent)
```
Comment (achristensen07) [inline Source/WebCore/page/RemoteFrame.cpp]: size_t.  This doesn't need to be able to support negative values.  Same for all the places where indentation is passed around.

File: Source/WebCore/page/RemoteFrameView.h [hunk 2]
```diff
@@ -36,6 +36,7 @@ class RemoteFrameView final : public FrameView {
     static Ref<RemoteFrameView> create(RemoteFrame& frame) { return adoptRef(*new RemoteFrameView(frame)); }
 
     Type viewType() const final { return Type::Remote; }
+    RemoteFrame& frame() const { return m_frame.get(); }
```
Comment (achristensen07) [inline Source/WebCore/page/RemoteFrameView.h]: Can we do something like this instead?

const RemoteFrame& frame() const { return m_frame.get(); }
RemoteFrame& frame() { return m_frame.get(); }

That makes it so if you have a const RemoteFrameView&, you can't mutate its frame.

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 3]
```diff
@@ -649,6 +651,11 @@ void write(TextStream& ts, const RenderObject& o, OptionSet<RenderAsTextFlag> be
                 if (RenderLayer* layer = root->layer())
                     writeLayers(ts, *layer, *layer, layer->rect(), behavior);
             }
+        } else if (is<RemoteFrameView>(widget)) {
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp]: Let's make this a pure virtual function on FrameView and have each type of FrameView implement its own functionality.  That way we don't have lots of places where we have the "if (local) else if (remote)" pattern.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp [hunk 4]
```diff
@@ -1358,6 +1358,18 @@ void WebProcessProxy::postMessageToRemote(WebCore::ProcessIdentifier destination
...
 }
 
+void WebProcessProxy::renderTreeDump(WebCore::ProcessIdentifier destinationProcessIdentifier, WebCore::FrameIdentifier frameIdentifier, int baseIndent, CompletionHandler<void(String)>&& completionHandler)
+{
+    String result;
+    auto webProcessProxy = processForIdentifier(destinationProcessIdentifier);
+    if (webProcessProxy) {
+        auto reply = webProcessProxy->sendSync(Messages::WebProcess::RenderTreeDump(frameIdentifier, baseIndent), 0);
+        if (reply)
+            std::tie(result) = reply.takeReply();
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebProcessProxy.cpp]: This use of std::tie is a little strange.  Can we just return completionHandler(reply.takeReply()) instead?

Also, early return completionHandler({ }) if webProcessProxy or reply aren't truthy to reduce indentation and keep code flow more straightforward.

File: Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp [hunk 5]
```diff
@@ -85,4 +85,13 @@ void WebRemoteFrameClient::changeLocation(WebCore::FrameLoadRequest&& request)
     });
 }
 
+String WebRemoteFrameClient::renderTreeDump(WebCore::ProcessIdentifier processIdentifier, WebCore::FrameIdentifier frameIdentifier, int baseIndent)
+{
+    String result;
+    auto reply = WebProcess::singleton().parentProcessConnection()->sendSync(Messages::WebProcessProxy::RenderTreeDump(processIdentifier, frameIdentifier, baseIndent), 0);
+    if (reply)
+        std::tie(result) = reply.takeReply();
```
Comment (achristensen07) [inline Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp]: ditto.
if (!reply)
    return completionHandler({ });

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 6]
```diff
@@ -917,9 +918,7 @@ void WebProcess::terminate()
 
 bool WebProcess::didReceiveSyncMessage(IPC::Connection& connection, IPC::Decoder& decoder, UniqueRef<IPC::Encoder>& replyEncoder)
 {
-    if (messageReceiverMap().dispatchSyncMessage(connection, decoder, replyEncoder))
-        return true;
-    return false;
+    return didReceiveSyncWebProcessMessage(connection, decoder, replyEncoder);
```
Comment (achristensen07) [inline Source/WebKit/WebProcess/WebProcess.cpp:923]: Does the generated code try passing the message to the message receiver map?

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 7]
```diff
@@ -1402,6 +1401,15 @@ void WebProcess::remotePostMessage(WebCore::FrameIdentifier identifier, std::opt
 
     domWindow->postMessageFromRemoteFrame(*globalObject, target, message);
 }
+
+void WebProcess::renderTreeDump(WebCore::FrameIdentifier frameIdentifier, int baseIndent, CompletionHandler<void(String)>&& completionHandler)
+{
+    auto* webFrame = WebProcess::singleton().webFrame(frameIdentifier);
+    ASSERT(webFrame);
```
Comment (achristensen07) [inline Source/WebKit/WebProcess/WebProcess.cpp]: ASSERT_NOT_REACHED, early return completionHandler({ });

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 8]
```diff
@@ -649,6 +651,11 @@ void write(TextStream& ts, const RenderObject& o, OptionSet<RenderAsTextFlag> be
...
+            RemoteFrameView& view = downcast<RemoteFrameView>(*widget);
+            auto& remoteFrame = view.frame();
+            ts << remoteFrame.renderTreeDump(ts.indent());
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp]: "renderTreeDump" probably needs a better name.  "renderTreeAsText"?

File: Source/WebCore/page/RemoteFrame.cpp (hunk 1 above)
Comment (chirags27) [reply Source/WebCore/page/RemoteFrame.cpp]: Yes, I used an int here mainly because `m_indent` is stored as an int and setIndent() takes in an int, but I can change that.

File: Source/WebKit/WebProcess/WebProcess.cpp (hunk 6 above)
Comment (chirags27) [reply Source/WebKit/WebProcess/WebProcess.cpp:923]: It directly does this:

```
bool WebProcess::didReceiveSyncWebProcessMessage(IPC::Connection& connection, IPC::Decoder& decoder, UniqueRef<IPC::Encoder>& replyEncoder)
{
    if (decoder.messageName() == Messages::WebProcess::RenderTreeDump::name())
        return IPC::handleMessageSynchronous<Messages::WebProcess::RenderTreeDump>(connection, decoder, replyEncoder, this, &WebProcess::renderTreeDump);
    UNUSED_PARAM(connection);
    UNUSED_PARAM(decoder);
    UNUSED_PARAM(replyEncoder);
#if ENABLE(IPC_TESTING_API)
    if (connection.ignoreInvalidMessageForTesting())
        return false;
#endif // ENABLE(IPC_TESTING_API)
    ASSERT_NOT_REACHED_WITH_MESSAGE("Unhandled synchronous message %s to %" PRIu64, description(decoder.messageName()), decoder.destinationID());
    return false;
}
```

I changed this because `WebPage::didReceiveSyncMessage` does something similar, but looks like this led to test failures. How is the `messageReceiverMap()` involved here?

File: Source/WebKit/UIProcess/WebProcessProxy.cpp (hunk 4 above)
Comment (chirags27) [reply Source/WebKit/UIProcess/WebProcessProxy.cpp]: `reply.takeReply()` returns a tuple, which is why I needed this. I'm seeing the same pattern at other places in case of sync messages, am I missing something?

File: Source/WebCore/rendering/RenderTreeAsText.cpp (hunk 3 above)
Comment (chirags27) [reply Source/WebCore/rendering/RenderTreeAsText.cpp]: that's fair, but I'm worried that it might change what we do for LocalFrameView currently because the only way would be for us to call externalRepresentation() LocalFrameView.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp [hunk 9]
```diff
@@ -1358,6 +1358,25 @@ void WebProcessProxy::postMessageToRemote(WebCore::ProcessIdentifier destination
         webProcessProxy->send(Messages::WebProcess::RemotePostMessage(identifier, target, message), 0);
 }
 
+void WebProcessProxy::renderTreeAsText(WebCore::ProcessIdentifier destinationProcessIdentifier, WebCore::FrameIdentifier frameIdentifier, size_t baseIndent, OptionSet<WebCore::RenderAsTextFlag> behavior, CompletionHandler<void(String)>&& completionHandler)
+{
+    (void)behavior;
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebProcessProxy.cpp]: Remove

File: Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in [hunk 10]
```diff
@@ -5172,3 +5172,19 @@ header: <WebCore/LandmarkTypeInterface.h>
     Nose,
 };
 #endif // GPU_PROCESS
+
+header: <WebCore/RenderTreeAsText.h>
+[OptionSet] enum class WebCore::RenderAsTextFlag : uint64_t {
```
Comment (achristensen07) [inline Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in]: : uint16_t

File: Source/WebCore/rendering/RenderTreeAsText.h [hunk 11]
```diff
@@ -54,7 +54,7 @@ enum class RenderAsTextFlag {
 };
 
 // You don't need pageWidthInPixels if you don't specify RenderAsTextInPrintingMode.
-WEBCORE_EXPORT String externalRepresentation(LocalFrame*, OptionSet<RenderAsTextFlag> = { });
+WEBCORE_EXPORT String externalRepresentation(LocalFrame*, OptionSet<RenderAsTextFlag> = { }, int indent = 0);
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.h]: size_t

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 12]
```diff
@@ -936,7 +933,7 @@ static void updateLayoutIgnoringPendingStylesheetsIncludingSubframes(Document& d
     }
 }
 
-String externalRepresentation(LocalFrame* frame, OptionSet<RenderAsTextFlag> behavior)
+String externalRepresentation(LocalFrame* frame, OptionSet<RenderAsTextFlag> behavior, int indent)
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp]: size_t

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 13]
```diff
@@ -908,9 +904,10 @@ static TextStream createTextStream(const Document& document)
     return { TextStream::LineMode::MultipleLine, formattingFlags() };
 }
 
-static String externalRepresentation(RenderBox& renderer, OptionSet<RenderAsTextFlag> behavior)
+static String externalRepresentation(RenderBox& renderer, OptionSet<RenderAsTextFlag> behavior, int indent)
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp]: size_t

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 14]
```diff
@@ -642,14 +644,8 @@ void write(TextStream& ts, const RenderObject& o, OptionSet<RenderAsTextFlag> be
...
         Widget* widget = downcast<RenderWidget>(o).widget();
-        if (is<LocalFrameView>(widget)) {
-            LocalFrameView& view = downcast<LocalFrameView>(*widget);
-            auto* localFrame = dynamicDowncast<LocalFrame>(view.frame());
-            if (RenderView* root = localFrame ? localFrame->contentRenderer() : nullptr) {
-                if (RenderLayer* layer = root->layer())
-                    writeLayers(ts, *layer, *layer, layer->rect(), behavior);
-            }
-        }
+        if (auto *frameView = dynamicDowncast<FrameView>(widget))
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp]: auto* frameView

File: Source/WebCore/css/CSSPrimitiveValue.cpp [hunk 15]
```diff
@@ -541,8 +541,13 @@ Ref<CSSPrimitiveValue> CSSPrimitiveValue::create(const Length& length, const Ren
     case LengthType::Fixed:
         return create(adjustFloatForAbsoluteZoom(length.value(), style), CSSUnitType::CSS_PX);
     case LengthType::Calculated:
-        // FIXME: Do we have a guarantee that CSSCalcValue::create will not return null here?
-        return create(CSSCalcValue::create(length.calculationValue(), style).releaseNonNull());
+    {
+        auto calcValue = CSSCalcValue::create(length.calculationValue(), style);
```
Comment (achristensen07) [inline Source/WebCore/css/CSSPrimitiveValue.cpp]: Remove unrelated changes.

File: Source/WebKit/UIProcess/WebProcessProxy.cpp [hunk 16]
```diff
@@ -1358,6 +1358,25 @@ void WebProcessProxy::postMessageToRemote(WebCore::ProcessIdentifier destination
...
+    auto webProcessProxy = processForIdentifier(destinationProcessIdentifier);
+    if (!webProcessProxy) {
+        completionHandler({ });
+        return;
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebProcessProxy.cpp]: These can be on one line.  Some people don't like that.  Either way is fine.

File: Source/WebKit/WebProcess/WebProcess.h [hunk 17]
```diff
@@ -47,6 +47,7 @@
 #include <WebCore/PageIdentifier.h>
 #include <WebCore/PluginData.h>
 #include <WebCore/RegistrableDomain.h>
+#include <WebCore/RenderTreeAsText.h>
```
Comment (achristensen07) [inline Source/WebKit/WebProcess/WebProcess.h]: Can the enum be forward declared instead of adding a new include to a header and slowing down the build?

File: Source/WebKit/UIProcess/WebProcessProxy.h [hunk 18]
```diff
@@ -47,6 +47,7 @@
 #include <WebCore/PageIdentifier.h>
 #include <WebCore/ProcessIdentifier.h>
 #include <WebCore/RegistrableDomain.h>
+#include <WebCore/RenderTreeAsText.h>
```
Comment (achristensen07) [inline Source/WebKit/UIProcess/WebProcessProxy.h]: Can the enum be forward declared?

File: Source/WebCore/page/FrameView.h [hunk 19]
```diff
@@ -26,13 +26,22 @@
 #pragma once
 
 #include "ScrollView.h"
+#include "RenderTreeAsText.h"
```
Comment (achristensen07) [inline Source/WebCore/page/FrameView.h]: Can the enum be forward declared?

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebCore/rendering/RenderTreeAsText.h [hunk 20]
```diff
@@ -38,24 +38,25 @@ class Element;
...
-    ShowCompositedLayers    = 1 << 2, // Show which layers are composited.
-    ShowOverflow            = 1 << 3, // Print layout and visual overflow.
-    ShowSVGGeometry         = 1 << 4, // Print additional geometry for SVG objects.
-    ShowLayerFragments      = 1 << 5, // Print additional info about fragmented RenderLayers
-    ShowAddresses           = 1 << 6, // Show layer and renderer addresses.
-    ShowIDAndClass          = 1 << 7, // Show id and class attributes
-    PrintingMode            = 1 << 8, // Dump the tree in printing mode.
-    DontUpdateLayout        = 1 << 9, // Don't update layout, to make it safe to call showLayerTree() from the debugger inside layout or painting code.
-    ShowLayoutState         = 1 << 10, // Print the various 'needs layout' bits on renderers.
+enum class RenderAsTextFlag : uint16_t {
```
Comment (chirags27) [inline Source/WebCore/rendering/RenderTreeAsText.h:41]: note: this should match RenderTreeExternalRepresentationFlags

File: Source/WebCore/platform/Widget.h [hunk 21]
```diff
@@ -128,7 +128,7 @@ class Widget : public RefCounted<Widget>, public CanMakeWeakPtr<Widget> {
 
     void setIsSelected(bool);
 
-    virtual bool isFrameView() const { return false; }
+    virtual bool isLocalFrameView() const { return false; }
```
Comment (chirags27) [inline Source/WebCore/platform/Widget.h:131]: I changed this to avoid confusion going forward

File: LayoutTests/http/tests/site-isolation/basic-iframe-render-output-expected.txt [hunk 22]
```diff
@@ -0,0 +1,13 @@
+layer at (0,0) size 800x600
```
Comment (achristensen07) [inline LayoutTests/http/tests/site-isolation/basic-iframe-render-output-expected.txt:1]: This needs to be skipped in LayoutTests/platform/mac-wk1/TestExpectations
Just skip the entire directory.  Site isolation will never be implemented for wk1.

File: Source/WebCore/platform/Widget.h (hunk 21 above)
Comment (achristensen07) [reply Source/WebCore/platform/Widget.h:131]: That's a good change.

File: Source/WebCore/page/RemoteFrame.h [hunk 23]
```diff
@@ -27,6 +27,7 @@
 
 #include "Frame.h"
 #include "LayerHostingContextIdentifier.h"
+#include "RenderTreeAsText.h"
```
Comment (achristensen07) [inline Source/WebCore/page/RemoteFrame.h]: Forward declare RenderAsTextFlag instead of this.  Otherwise it's unnecessarily adding to the build time to include a header from another header.

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 24]
```diff
@@ -642,13 +644,9 @@ void write(TextStream& ts, const RenderObject& o, OptionSet<RenderAsTextFlag> be
...
     if (is<RenderWidget>(o)) {
         Widget* widget = downcast<RenderWidget>(o).widget();
-        if (is<LocalFrameView>(widget)) {
-            LocalFrameView& view = downcast<LocalFrameView>(*widget);
-            auto* localFrame = dynamicDowncast<LocalFrame>(view.frame());
-            if (RenderView* root = localFrame ? localFrame->contentRenderer() : nullptr) {
-                if (RenderLayer* layer = root->layer())
-                    writeLayers(ts, *layer, *layer, layer->rect(), behavior);
-            }
+        if (widget) {
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp:647]: This check is unnecessary.  dynamicDowncast will return null if given null.

File: Source/WebCore/page/LocalFrameView.cpp [hunk 25]
```diff
@@ -6330,6 +6331,14 @@ float LocalFrameView::deviceScaleFactor() const
     return 1;
 }
 
+void LocalFrameView::writeRenderTreeAsText(TextStream& ts, OptionSet<RenderAsTextFlag> behavior)
+{
+    auto* localFrame = dynamicDowncast<LocalFrame>(frame());
+    if (!localFrame)
+        return;
+    ts << externalRepresentationForLocalFrame(localFrame, behavior, ts.indent());
```
Comment (achristensen07) [inline Source/WebCore/page/LocalFrameView.cpp]: This adds an unnecessary String allocation that wasn't there before.  Can we pass ts as a parameter to the common function, then create a TextStream in WebProcess::renderTreeAsText?

File: Source/WebCore/rendering/RenderTreeAsText.cpp [hunk 26]
```diff
@@ -955,6 +953,26 @@ String externalRepresentation(LocalFrame* frame, OptionSet<RenderAsTextFlag> beh
     return externalRepresentation(*renderer, behavior);
 }
 
+String externalRepresentationForLocalFrame(LocalFrame* frame, OptionSet<RenderAsTextFlag> behavior, size_t indent)
+{
+    ASSERT(frame);
```
Comment (achristensen07) [inline Source/WebCore/rendering/RenderTreeAsText.cpp]: This is an old anti pattern.  Pass in a LocalFrame& as a parameter instead.

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebCore/rendering/RenderTreeAsText.cpp (hunk 24 above)
Comment (chirags27) [reply Source/WebCore/rendering/RenderTreeAsText.cpp:647]: The check is for widget which is `downcast<RenderWidget>(o).widget();` , which I believe can still be null?

**Review by achristensen07 — APPROVED** [review]
